pkg/provider/vault/auth_iam.go

```
@@ -183,3 +143,87 @@ func (c *client) requestTokenWithIamAuth(ctx context.Context, iamAuth *esv1.Vaul
+		logger.V(1).Info("using Pod Identity for authentication")
+		// Return nil to let AWS SDK v1 container credential provider handle Pod Identity automatically
+		return nil, nil
+	}
+
+	// No IRSA or Pod Identity found.
+	return nil, errors.New(errNoAWSAuthMethodFound)
+}
+
+func (c *client) getCredsFromIRSAToken(ctx context.Context, tokenFile, region string, k kclient.Client, jwtProvider util.JwtProviderFactory) (*credentials.Credentials, error) {
+	// IRSA enabled service account, let's check that the jwt token filemount and file exists
+	if _, err := os.Stat(filepath.Clean(tokenFile)); err != nil {
+		return nil, fmt.Errorf(errIrsaTokenFileNotFoundOnPod, tokenFile, err)
+	}
+
+	// everything looks good so far, let's fetch the jwt token from AWS_WEB_IDENTITY_TOKEN_FILE
+	jwtByte, err := os.ReadFile(filepath.Clean(tokenFile))
+	if err != nil {
+		return nil, fmt.Errorf(errIrsaTokenFileNotReadable, tokenFile, err)
+	}
+
+	// let's parse the jwt token
+	parser := jwt.NewParser(jwt.WithoutClaimsValidation())
+
+	token, _, err := parser.ParseUnverified(string(jwtByte), jwt.MapClaims{})
+	if err != nil {
+		return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err) // JWT token parser error
+	}
+
+	var ns string
+	var sa string
+
+	// let's fetch the namespace and serviceaccount from parsed jwt token
+	if claims, ok := token.Claims.(jwt.MapClaims); ok {
+		k8s, ok := claims["kubernetes.io"].(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		ns, ok = k8s["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		saMap, ok := k8s["serviceaccount"].(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		sa, ok = saMap["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+	} else {
+		return nil, fmt.Errorf(errPodInfoNotFoundOnToken, tokenFile, err)
+	}
```

The `err` here is nil since it's already checked on line 199.

---

I'm not entirely sure which error are you referring to. These changes kind of cascaded beyond the scope of what I was working on, because all I did was move this code. Happy to improve it whilst we are touching it though.

I believe once the err is nil for parseUnverified its  we need to return an error if the claims are invalid right, I have adjusted the errors slightly to return an invalid claims error for all those cases.

let me know if that will suffice or if more changes are needed.

---

I'm saying that here:

```go
	token, _, err := parser.ParseUnverified(string(jwtByte), jwt.MapClaims{})
	if err != nil {
		return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err) // JWT token parser error
	}
```

There was a nil check for the error in these:

```go
return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
```

So in this fmt.Errorf call the error will always be just `nil`. So might as well just drop it. :)

---

ow right okay sorry yeah that is my bad lol. That I will fix